src/gj.rs

```diff
@@ -620,6 +620,78 @@ impl EGraph {
         }
     }
 
+    fn gj_for_atom<F>(
+        &self,
+        atom_i: usize,
+        timestamp_ranges: &[Range<u32>],
+        cq: &CompiledQuery,
+        mut f: F,
+    ) where
+        F: FnMut(&[Value]) -> Result,
+    {
+        let atom = &cq.query.atoms[atom_i];
+        // do the gj
+        if let Some((mut ctx, program, cols)) = Context::new(self, cq, timestamp_ranges) {
+            let start = Instant::now();
+            log::debug!(
+                        "Query:\n{q}\nNew atom: {atom}\nTuple: {tuple}\nJoin order: {order}\nProgram\n{program}",
```

Comment:
nits: `atom`/`atom_i` only makes sense for semi-naive. Moreover, if not doing semi-naive, `New atom: ...` does not make sense. It would be great if you could do a favor by also refactoring this part.

Reply:
done!